internal/utils/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,73 @@
+package ringbuffer
+
+type RingBuffer[T any] struct {
+	ring             []T
+	headPos, tailPos int
+	full             bool
+}
+
+func (r *RingBuffer[T]) Init(size int) {
+	r.ring = make([]T, size)
+}
+
+func (r *RingBuffer[T]) Len() int {
+	if r.full {
+		return len(r.ring)
+	}
+	if r.tailPos >= r.headPos {
+		return r.tailPos - r.headPos
+	}
+	return r.tailPos - r.headPos + len(r.ring)
+}
+
+func (r *RingBuffer[T]) PushBack(t T) {
+	if r.full || len(r.ring) == 0 {
+		r.grow()
+	}
+	r.ring[r.tailPos] = t
+	r.tailPos++
+	if r.tailPos == len(r.ring) {
+		r.tailPos = 0
+	}
+	if r.tailPos == r.headPos {
+		r.full = true
+	}
+}
+
+func (r *RingBuffer[T]) PopFront() (t T, ok bool) {
+	if r.Len() == 0 {
+		return
+	}
+	ok = true
+	r.full = false
+	t = r.ring[r.headPos]
+	var zeroValue T
+	r.ring[r.headPos] = zeroValue
+	r.headPos++
+	if r.headPos == len(r.ring) {
+		r.headPos = 0
+	}
+	return
```

Comment:
```suggestion
	return t, true
```

Reply:
We should panic here since pop from an empty queue must be a bug.